executor/adapter.go

```diff
@@ -307,8 +416,89 @@ func (a *ExecStmt) handleNoDelayExecutor(ctx context.Context, sctx sessionctx.Co
 	if err != nil {
 		return nil, err
 	}
+	return nil, err
+}
+
+func (a *ExecStmt) handlePessimisticDML(ctx context.Context, sctx sessionctx.Context, e Executor) error {
+	txn, err := sctx.Txn(false)
+	if err != nil {
+		return err
+	}
+	txnCtx := sctx.GetSessionVars().TxnCtx
+	retryCnt := uint(0)
+	for {
+		_, err = a.handleNoDelayExecutor(ctx, sctx, e)
+		if err != nil {
+			return err
+		}
+		p := txn.(pessimisticTxn)
+		keys, err1 := p.KeysNeedToLock()
+		if err1 != nil {
+			return err1
+		}
+		if len(keys) == 0 {
+			return nil
+		}
+		forUpdateTS := txnCtx.GetForUpdateTS()
+		err = txn.LockKeys(ctx, forUpdateTS, keys...)
+		if err == nil || !strings.Contains(err.Error(), tidbutil.WriteConflictMarker) {
+			return err
+		}
+		if retryCnt == config.GetGlobalConfig().PessimisticTxn.MaxRetryCount {
+			return errors.New("pessimistic lock retry limit reached")
+		}
+		retryCnt++
+		conflictTS := extractConflictTS(err.Error())
+		log.Info("pessimistic write conflict, retry statement",
+			zap.Uint64("txn", txn.StartTS()),
+			zap.Uint64("forUpdateTS", forUpdateTS),
+			zap.Uint64("conflictTS", conflictTS))
+		if conflictTS > forUpdateTS {
+			txnCtx.SetForUpdateTS(conflictTS)
+		} else {
+			ts, err1 := sctx.GetStore().GetOracle().GetTimestamp(ctx)
+			if err1 != nil {
+				return err1
+			}
+			txnCtx.SetForUpdateTS(ts)
+		}
+		e, err = a.buildExecutor(sctx)
+		if err != nil {
+			return err
+		}
+
+		// Rollback the statement change before retry it.
+		sctx.StmtRollback()
+		sctx.GetSessionVars().StmtCtx.ResetForRetry()
+
+		if err = e.Open(ctx); err != nil {
+			return err
+		}
+	}
+}
+
+func extractConflictTS(errStr string) uint64 {
+	strs := strings.Split(errStr, "conflictTS=")
```

Comment:
It's not easy to write the test because the code formats the error message is in 'tikv' package.
We can log a warning when failed to extract conflictTS.